src/SensitiveFloat.jl

```diff
@@ -0,0 +1,275 @@
+
+export multiply_sfs!, add_scaled_sfs!, combine_sfs!, add_sources_sf!
+
+@doc """
+A function value and its derivative with respect to its arguments.
+
+Attributes:
+  v:  The value
+  d:  The derivative with respect to each variable in
+      P-dimensional VariationalParams for each of S celestial objects
+      in a local_P x local_S matrix.
+  h:  The second derivative with respect to each variational parameter,
+      in the same format as d.  This is used for the full Hessian
+      with respect to all the sources.
+  hs: An array of per-source Hessians.  This will generally be reserved
+      for the Hessian of brightness values that depend only on one source.
+""" ->
+type SensitiveFloat{ParamType <: CelesteTypes.ParamSet, NumType <: Number}
+    # Actually a single value, but an Array to avoid memory allocation
+    v::Vector{NumType}
```

Comment:
I found when looking at memory allocation for the full Hessian that memory
was allocated each time the value of a SensitiveFloat was changed and (in
some cases at least) even accessed.  I don't know enough about the julia
compiler to account for this, but when I switched to using an array for the
value the memory allocation disappeared.  So when I say "to avoid memory
allocation" I mean "to avoid memory allocation whenever the value is used
or changed" rather than to avoid memory allocation up front.

On Tue, Mar 8, 2016 at 10:51 AM, Kyle Barbary <notifications@github.com>
wrote:

> In src/SensitiveFloat.jl
> <https://github.com/jeff-regier/Celeste.jl/pull/137#discussion_r55407321>:
>
> > +A function value and its derivative with respect to its arguments.
> > +
> > +Attributes:
> > +  v:  The value
> > +  d:  The derivative with respect to each variable in
> > +      P-dimensional VariationalParams for each of S celestial objects
> > +      in a local_P x local_S matrix.
> > +  h:  The second derivative with respect to each variational parameter,
> > +      in the same format as d.  This is used for the full Hessian
> > +      with respect to all the sources.
> > +  hs: An array of per-source Hessians.  This will generally be reserved
> > +      for the Hessian of brightness values that depend only on one source.
> > +""" ->
> > +type SensitiveFloat{ParamType <: CelesteTypes.ParamSet, NumType <: Number}
> > +    # Actually a single value, but an Array to avoid memory allocation
> > +    v::Vector{NumType}
>
> While we're looking at SensitiveFloat, I wanted to ask why v is an array
> instead of just v::Numtype here. The effect of using an array is opposite
> of what the comment implies.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jeff-regier/Celeste.jl/pull/137/files#r55407321>.
>